srml/grandpa/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+			use codec::{Encode, KeyedVec};
+
+			let auth_count = config.authorities.len() as u32;
+			config.authorities.iter().enumerate().for_each(|(i, v)| {
+				storage.insert((i as u32).to_keyed_vec(
+					::fg_primitives::well_known_keys::AUTHORITY_PREFIX),
+					v.encode()
+				);
+			});
+			storage.insert(
+				::fg_primitives::well_known_keys::AUTHORITY_COUNT.to_vec(),
+				auth_count.encode(),
+			);
+		});
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Report some misbehaviour.
+		fn report_misbehavior(origin, _report: Vec<u8>) -> Result {
+			ensure_signed(origin)?;
+			// TODO: https://github.com/paritytech/substrate/issues/1112
+			Ok(())
+		}
+
+		fn on_finalise(block_number: T::BlockNumber) {
+			if let Some(pending_change) = <PendingChange<T>>::get() {
+				if block_number == pending_change.scheduled_at {
+					Self::deposit_log(RawLog::AuthoritiesChangeSignal(
+						pending_change.delay,
+						pending_change.next_authorities.clone(),
+					));
+				}
+
+				if block_number == pending_change.scheduled_at + pending_change.delay {
+					Self::deposit_event(
+						RawEvent::NewAuthorities(pending_change.next_authorities.clone())
+					);
+					<AuthorityStorageVec<T::SessionKey>>::set_items(pending_change.next_authorities);
+					<PendingChange<T>>::kill();
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get the current set of authorities, along with their respective weights.
+	pub fn grandpa_authorities() -> Vec<(T::SessionKey, u64)> {
+		<AuthorityStorageVec<T::SessionKey>>::items()
+	}
+
+	/// Schedule a change in the authorities.
+	///
+	/// The change will be applied at the end of execution of the block
+	/// `in_blocks` after the current block. This value may be 0, in which
+	/// case the change is applied at the end of the current block.
+	///
+	/// No change should be signalled while any change is pending. Returns
+	/// an error if a change is already pending.
+	pub fn schedule_change(
+		next_authorities: Vec<(T::SessionKey, u64)>,
+		in_blocks: T::BlockNumber,
+	) -> Result {
+		if Self::pending_change().is_none() {
+			let scheduled_at = system::ChainContext::<T>::default().current_height();
+			<PendingChange<T>>::put(StoredPendingChange {
+				delay: in_blocks,
+				scheduled_at,
+				next_authorities,
+			});
+
+			Ok(())
+		} else {
+			Err("Attempt to signal GRANDPA change with one already pending.")
+		}
+	}
+
+	/// Deposit one of this module's logs.
+	fn deposit_log(log: Log<T>) {
+		<system::Module<T>>::deposit_log(<T as Trait>::Log::from(log).into());
+	}
+}
+
+impl<T: Trait> Module<T>
+where
+	AuthorityId: core::convert::From<<T as Trait>::SessionKey>
+{
+	/// See if the digest contains any scheduled change.
+	pub fn scrape_digest_change(log: &Log<T>)
+		-> Option<ScheduledChange<T::BlockNumber>>
+	{
+		<Log<T> as GrandpaChangeSignal<T::BlockNumber>>::as_signal(log)
```

Comment:
does this really need to be a Log? why not just have it as a transient storage item?

Reply:
it has to be a log or there is no way to follow consensus changes with only the header chain

Reply:
if it were a transient storage item, light clients would basically be forked out

Reply:
and a state-proof would be too heavyweight?

Reply:
We would still need to have some kind of signal _in the header chain_ because that is what light clients follow. Otherwise they have no way of knowing if they've missed a transition proof, or where to find that in the chain. We can't rely on peers to send us the right data in this case because it is critical for consensus.

i won't rule out coupling state-proofs with a lightweight signal in order to avoid header-chain bloat. In general I'd like to classify GRANDPA as unstable for now and there are bound to be more changes before it becomes stable as we add features like dynamic-set fallback and commit messages on-chain.

This is a minimal integration and by no means final -- more like "good enough for a testnet".

Reply:
fair enough